### PR TITLE
Replaces the static field with AsyncLocal and simplifies the connection mgmt

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Connects to Particular/NServiceBus.Testing#29

There were some problems with this version of multi tenant DB. I've noticed that it does not go to the correct (tenant DB), at least for some of the DB calls. I've done two changes
 * The incoming pipeline behavior now only sets the correct connection string (does not open the connection). This is because the new connection management in V7 of NH does a better job than in V5
 * The mentioned behavior has been moved forward to the transport receive stage to execute before the outbox is involved
 * The static storage used to pass context between the behavior and the NH connection provider has been changed to use .NET 4.6 `AsyncLocal<T>`

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@SimonCropp @HEskandari please review.

File: samples/outbox/multi-tenant/Version_6/Receiver/Receiver.csproj [hunk 1]
```diff
@@ -9,7 +9,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>Receiver</RootNamespace>
     <AssemblyName>Receiver</AssemblyName>
-    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
```
Comment (SimonCropp) [inline samples/outbox/multi-tenant/Version_6/Receiver/Receiver.csproj]: why 4.6?

File: samples/outbox/multi-tenant/Version_6/Sender/app.config [hunk 2]
```diff
@@ -19,4 +19,12 @@
       </providers>
     </roleManager>
   </system.web>
+  <runtime>
+    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+      <dependentAssembly>
+        <assemblyIdentity name="NServiceBus.Core" publicKeyToken="9fc386479f8a226c" culture="neutral" />
```
Comment (SimonCropp) [inline samples/outbox/multi-tenant/Version_6/Sender/app.config]: why is this required?

Comment (SzymonPobiega) [inline samples/outbox/multi-tenant/Version_6/Sender/app.config]: It isn't:( package manager added it silently...

File: samples/outbox/multi-tenant/Version_6/Receiver/Receiver.csproj (hunk 1 above)
Comment (SzymonPobiega) [inline samples/outbox/multi-tenant/Version_6/Receiver/Receiver.csproj]: `AsyncLocal<T>` is 4.6. I think it is fair to require this since this is a very specific sample. Previous .NET editions did not have a good solution for async local state according to @danielmarbach

File: samples/outbox/multi-tenant/Version_6/Sender/app.config (hunk 2 above)
Comment (SimonCropp) [inline samples/outbox/multi-tenant/Version_6/Sender/app.config]: yeah i have started doing many of my nuget updates using sublime text search and replace

File: samples/outbox/multi-tenant/Version_6/Receiver/Receiver.csproj (hunk 1 above)
Comment (SimonCropp) [inline samples/outbox/multi-tenant/Version_6/Receiver/Receiver.csproj]: ahh ok then. this is fine. i suspect the other samples will go 4.6 within the next few months anyway

File: samples/outbox/multi-tenant/Version_6/Receiver/ExtractTenantConnectionStringBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Configuration;
+using System.Data;
```
Comment (SimonCropp) [inline samples/outbox/multi-tenant/Version_6/Receiver/ExtractTenantConnectionStringBehavior.cs]: redundant usings
